include/dali/core/mm/monotonic_resource.h

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_MM_MONOTONIC_RESOURCE_H_
+#define DALI_CORE_MM_MONOTONIC_RESOURCE_H_
+
+#include <cassert>
+#include <algorithm>
+#include <utility>
+#include "dali/core/mm/mr.h"
+#include "dali/core/mm/detail/util.h"
+#include "dali/core/mm/detail/align.h"
+#include "dali/core/small_vector.h"
+
+namespace dali {
+namespace mm {
+
+/**
+ * @brief Monotonic resource which returns memory from a preallocated buffer.
+ *
+ * Monotonic buffer resources don't require manual deallocation of the returned pointers.
+ * The memory is freed in bulk when the underlying buffer is freed.
+ */
+class monotonic_buffer_resource : public memory_resource {
+ public:
+  monotonic_buffer_resource() = default;
+  monotonic_buffer_resource(void *memory, size_t bytes)
+  : curr_(static_cast<char*>(memory)), limit_(static_cast<char*>(memory) + bytes) {}
+
+  size_t avail(size_t alignment) const {
+    return limit_ - detail::align_ptr(curr_, alignment);
+  }
+
+ protected:
+  void *do_allocate(size_t bytes, size_t alignment) override {
+    char *ret = detail::align_ptr(curr_, alignment);
+    if (ret + bytes > limit_)
+      throw std::bad_alloc();
+    curr_ = ret + bytes;
+    return ret;
+  }
+
+  // don't deallocate at all
+  void do_deallocate(void *data, size_t bytes, size_t alignment) override {
+  }
+
+  char *curr_ = nullptr, *limit_ = nullptr;
+};
+
+/**
+ * @brief Monotonic resource which allocates memory from an upstream resource, storing
+ *        the metadata in the same memory blocks as the allocated buffers.
+ *
+ * Monotonic resources don't require manual deallocation of memory.
+ * The lifetime of a monotnic resource is limited and all memory will be deallocated in bulk
+ * when the resource is destroyed.
+ */
+class monotonic_host_resource : public memory_resource {
+ public:
+  explicit monotonic_host_resource(memory_resource *upstream,
+                                   size_t next_block_size = 1024)
+  : upstream_(upstream), next_block_size_(next_block_size) {}
+
+  ~monotonic_host_resource() {
```

Comment:
```suggestion
  ~monotonic_host_resource() override {
```

Reply:
`override` keyword was added to make it an error when you _think_ that you override but you don't. There's no such risk with destructors.

Reply:
> override also verifies, that the superclass's function is declared virtual. 
But we don't care. It's not important. It doesn't give anything or even doesn't save us from errors. A class can very well have an implicit non-trivial destructor which would not be used if the base class pointer was deleted.
Initally i made the destructors of the base classes `protected` and non-virtual, but changed it to public virtual _only_ to match `std::pmr::memory_resource`.